File: languages/go/idiomatic/brainfuck.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
)

const MaxCellSize = 30000

type State struct {
	cells [MaxCellSize]int
	pos   int
}

func main() {
	if len(os.Args) != 2 {
		errorExit("You must provide a filename as an argument")
	}

	program, err := ioutil.ReadFile(os.Args[1])
	if err != nil {
		errorExit("Could not open file")
	}

	state := &State{}
	instrPointer := 0

	for instrPointer < len(program) {
		command := program[instrPointer]

		switch command {
		case '+':
			state.writeCurCell(state.readCurCell() + 1)
		case '-':
			state.writeCurCell(state.readCurCell() - 1)
		case '>':
			state.pos++
		case '<':
			if state.pos == 0 {
				errorExit("Attempted to decrement data pointer below 0")
			}
			state.pos--
		case '.':
			fmt.Printf("%c", state.readCurCell())
		case ',':
			var value int
			fmt.Scanf("%c", &value)
			state.writeCurCell(value)
		case '[':
			if state.readCurCell() == 0 {
				depth := 1
				for depth > 0 && instrPointer < len(program) {
					instrPointer++
					if program[instrPointer] == '[' {
						depth++
					} else if program[instrPointer] == ']' {
						depth--
					}
				}
			}
		case ']':
			if state.readCurCell() != 0 {
				depth := 1
				for depth > 0 && instrPointer >= 0 {
					instrPointer--
					if program[instrPointer] == ']' {
						depth++
					} else if program[instrPointer] == '[' {
						depth--
					}
				}
			}
		}
		instrPointer++
	}
}

func (state *State) readCurCell() int {
	if state.pos >= len(state.cells) {
		return 0
	} else {
		return state.cells[state.pos]
	}
}

func (state *State) writeCurCell(value int) {
	state.cells[state.pos] = value
}

func errorExit(msg string) {
	fmt.Fprintln(os.Stderr, msg)
	os.Exit(1)
}

File: languages/swift/idiomatic/brainfuck.swift

import Foundation

let MAX_CELL_SIZE = 30000

struct State {
    var cells: [UInt8] = Array(repeating: 0, count: MAX_CELL_SIZE)
    var pos: Int = 0
}

func error(_ msg: String) -> Never {
    print(msg)
    exit(1)
}

func readCurCell(_ state: inout State) -> UInt8 {
    return state.cells[state.pos]
}

func writeCurCell(_ state: inout State, value: UInt8) {
    state.cells[state.pos] = value
}

func incCurCell(_ state: inout State) {
    writeCurCell(&state, value: readCurCell(&state) &+ 1)
}

func decCurCell(_ state: inout State) {
    writeCurCell(&state, value: readCurCell(&state) &- 1)
}

func main(args: [String]) {
    let argc = args.count

    if argc != 2 {
        error("You must provide a filename as an argument")
    }

    let fileURL = URL(fileURLWithPath: args[1])

    guard let program_str = try? String(contentsOf: fileURL) else {
        error("Could not open file")
    }

	let program = [UInt8](program_str.utf8);

    var state = State()
    var instr_pointer = 0

    while instr_pointer < program.count {
        let command = program[instr_pointer]

        switch UnicodeScalar(command) {
        case "+":
            incCurCell(&state)
        case "-":
            decCurCell(&state)
        case ">":
            state.pos += 1
        case "<":
            if state.pos == 0 {
                error("Attempted to decrement data pointer below 0")
            }
            state.pos -= 1
        case ".":
            print(UnicodeScalar(readCurCell(&state)), terminator: "")
        case ",":
            writeCurCell(&state, value: UInt8(getchar()))
        case "[":
            if readCurCell(&state) == 0 {
                var depth = 1
                while depth > 0 && instr_pointer < program.count {
                    instr_pointer += 1
                    if UnicodeScalar(program[instr_pointer]) == "[" {
                        depth += 1
                    } else if UnicodeScalar(program[instr_pointer]) == "]" {
                        depth -= 1
                    }
                }
            }
        case "]":
            if readCurCell(&state) != 0 {
                var depth = 1
                while depth > 0 && instr_pointer >= 0 {
                    instr_pointer -= 1
                    if UnicodeScalar(program[instr_pointer]) == "]" {
                        depth += 1
                    } else if UnicodeScalar(program[instr_pointer]) == "[" {
                        depth -= 1
                    }
                }
            }
        default:
            break
        }
        instr_pointer += 1
    }
}

let arguments = CommandLine.arguments
main(args: arguments)
